### PR TITLE
ARCHIE-210: Add support for mode GPT-4 plus some other odds and ends

### PR DESCRIPTION
**Pull request by cgianelle**
https://cipherhealth.atlassian.net/browse/ARCHIE-210
https://cipherhealth.atlassian.net/browse/ARCHIE-211
https://cipherhealth.atlassian.net/browse/ARCHIE-212
https://cipherhealth.atlassian.net/browse/ARCHIE-177

### CODE REVIEW

File: Dockerfile.websockets [hunk 1]
```diff
@@ -4,4 +4,5 @@ COPY requirements.txt .
 RUN pip install --no-cache-dir -r requirements.txt
 COPY . .
 EXPOSE 8080
-CMD ["python", "slack_gpt_bot.py"]
\ No newline at end of file
+# CMD ["python", "slack_gpt_bot.py"]
```
Comment (markusos) [inline Dockerfile.websockets:7]: Do we want to clean this up?

File: slack_gpt_bot.py [hunk 2]
```diff
@@ -1,218 +1,283 @@
...
+            return (OPENAI_MODEL_DEFAULT, OPENAI_MODEL_MAX_TOKENS)
+
+    def determine_openai_model_4_to_use(self, input_token_count):
+        if OPENAI_MODEL_4_EXTENDED_FEATURE_FLAG:
+            if input_token_count > OPENAI_MODEL_4_CROSSOVER_POINT:
+                return (OPENAI_MODEL_4_EXTENDED, OPENAI_MODEL_4_EXTENDED_MAX_TOKENS)
+            else:
+                return (OPENAI_MODEL_4_DEFAULT, OPENAI_MODEL_4_MAX_TOKENS)
+        else:
+            return (OPENAI_MODEL_4_DEFAULT, OPENAI_MODEL_4_MAX_TOKENS)
```
Comment (markusos) [inline slack_gpt_bot.py:114]: Maybe we can avoid the nested `if` here:
```
use_extended_model = input_token_count > OPENAI_MODEL_4_CROSSOVER_POINT
if OPENAI_MODEL_4_EXTENDED_FEATURE_FLAG and use_extended_model:
    return (OPENAI_MODEL_4_EXTENDED, OPENAI_MODEL_4_EXTENDED_MAX_TOKENS)
else:
    return (OPENAI_MODEL_4_DEFAULT, OPENAI_MODEL_4_MAX_TOKENS)
end
````

File: slack_gpt_bot.py [hunk 3]
```diff
@@ -1,218 +1,283 @@
...
+            were denied access.
+            
+            if channel_id != 'C057NBLL2G4': #lock to test channel for beta
+                slack_resp = app.client.chat_postMessage( 
+                    channel=channel_id,
+                    thread_ts=thread_ts,
+                    text="Our apologies, however the Beta ChatGPT bot is not allowed outside of the beta-slack-chatgpt-bot channel"
+                )
+                return
+            '''
```
Comment (markusos) [inline slack_gpt_bot.py:168]: Maybe we can leave this in place but have a feature flag to disable it, just like for OPENAI_MODEL_4_EXTENDED_FEATURE_FLAG?

**Review by markusos — APPROVED** [review]
Overall looks good, left some thoughts. Maybe we can clean up some of the commented out code and outdated comments

**Review by chuchothecat — APPROVED** [review]
LGTM.  Agreed it could benefit from some minor cleanup, seems good to go though for all practical purposes.

**Comment by cgianelle** [comment]
@markusos @chuchothecat Thanks for the feedback - I'll create stories and address in the next revision

**Pull request merged**